server/docs/design/block-verification.md

```diff
@@ -0,0 +1,198 @@
+    end
+    
+    else (3) Append more Block Items
+    loop
+    V->>S: addBlockItems(items)
+    V-->>U: return without blocking
+    rect rgb(230, 230, 230)
+    S->>S: *Continues hash computation asynchronously
+    end
+    end
+
+    else (4) Append BlockItems with block_proof
+   
+    V->>S: addBlockItems(items with block_proof)
+    V-->>U: return without blocking    
+    S->>S: async completeHashing()   
+
+    S->>SV: (5) verifySignature(signature, computedHash, blockNumber)    
+
+    Note over SV,BSM: (6) Compare computed hash and signature
+    alt (7) Verification Fails
+      SV->>BSM: updateBlockStatus(blockNumber, ERROR, SIGNATURE_INVALID      
+    else (8) Verification Succeeds
+      SV->>BSM: updateBlockStatus(blockNumber, VERIFIED, NONE)      
+    end
+    Note over BSM: (9) Follow-up to downstream services
+    end
+
+```
+**Note:** The gray boxes on async hash computation will be implemented later only if we need to improve the latency of the verified blocks stream. (see note at the end of the document)
+## Interfaces
+
+### VerificationHandler
+
+```java
+public interface VerificationHandler {
+  void onBlockItemsReceived(List<BlockItem> blockItems);  
+}
+```
+### BlockHashingSessionFactory
+
+```java
+import java.util.concurrent.ExecutorService;
+
+public interface BlockHashingSessionFactory {
+  BlockHashingSession createSession(List<BlockItem> initialBlockItems, ExecutorService executorService, SignatureVerifier signatureVerifier);
+}
+```
+### BlockHashingSession
+```java
+/* Once hashing is completed internally, calls SignatureVerifier to continue with verification process */
+public interface BlockHashingSession {
+  void addBlockItem(BlockItem item);
+  CompletableFuture<Void> completeHashing(); // triggers final hash computation asynchronously  
+}
+```
+### SignatureVerifier
+```java
+public interface SignatureVerifier {
+  void verifySignature(byte[] signature, byte[] computedHash, long blockNumber);  
+}
+```
+### BlockStatusManager
+```java
+
+public interface BlockStatusManager {
+    void updateBlockStatus(long blockNumber, BlockVerificationStatus status, VerificationError error);
+}
+```
+
+## Enums
+```java
+public enum BlockVerificationStatus {
+  VERIFIED,
+  ERROR
+}
+
+public enum VerificationError {
+  NONE,
+  SIGNATURE_INVALID,
+  SYSTEM_ERROR
+}
+```
+
+## Metrics
+
+- **blocks_received**: Counter of the number of blocks received for verification
+- **blocks_verified**: Counter of the number of blocks verified
+- **blocks_unverified**: Counter of the number of blocks that have not been verified, is the difference between blocks_received and blocks_verified.
+- **blocks_signature_invalid**: Counter of the number of blocks with invalid signatures
+- **blocks_system_error**: Counter of the number of blocks with system errors
+- **block_verification_time**: Histogram of the time taken to verify a block, gives the node operator an idea of the time taken to verify a block.
+- 
+## Exceptions
```

Comment:
Are these exceptions or response codes?

Reply:
I think we can further refine this, due to the Async Nature of verification handler, exceptions becomes response codes or statuses of the work done. So I would say that in this case they are both.